src/dbt_score/rule.py

```diff
@@ -78,6 +81,25 @@ def __init_subclass__(cls, **kwargs) -> None:  # type: ignore
         if not hasattr(cls, "description"):
             raise AttributeError("Subclass must define class attribute `description`.")
 
+        cls.resource_type = cls._introspect_resource_type()
+
+    @classmethod
+    def _introspect_resource_type(cls) -> Type[Evaluable]:
+        evaluate_func = getattr(cls, "_orig_evaluate", cls.evaluate)
+
+        sig = inspect.signature(evaluate_func)
+        resource_type_argument = first_true(
+            sig.parameters.values(),
+            pred=lambda arg: arg.annotation in typing.get_args(Evaluable),
+        )
+
+        if not resource_type_argument:
+            raise TypeError(
+                "Subclass must implement method `evaluate` with an annotated Model or Source argument."
+            )
```

Comment:
Since the type-annotation is critical to a proper resource-type discriminator being set on the class, I thought that it was valid to throw an error here if there isn't a param typed as `Source` or `Model`.